### ISSUE TITLE
A wrapper for docker-py

### ISSUE THREAD

**Issue by cecton** [aio-libs/aiodocker#98]
Before I started working on this project, I had my own Docker client library for aiohttp;

https://github.com/tenforce/docker-py-aiohttp

But I took a very different approach. My idea was to make docker-py return futures instead of result directly. It took me a very long time before I finally figured out a solution but recently, I succeeded. I made some kind of wrapper library around docker-py that will re-use most of the code of docker-py but redirect the HTTP calls to aiohttp instead and return futures instead of directly the result. I only had to re-write the json stream parser and underlying helpers like that. At this point, the entire low-level API of docker-py is working and the high-level one will need a little bit more tweak.

Now, to be honest, I'm struggling between aiodocker and aiodockerpy. Both have pros and cons:

* For example, the code of aiodocker is clean and brand new while aiodockerpy is more hacking around the requests handlers.
* On the other hand, aiodockerpy doesn't need the same kind of maintenance: the API changes is actually handled by docker-py itself, not by aiodockerpy. I get almost everything "for free"... even the documentation is the same.

I would really like to heard your thoughts about all of this.

**Comment by asvetlov** [comment]
Pros and cons are obvious.
Honestly I prefer current solution.
Development of libraries built on top sync basement is a pain, take a look on aiomysql or aiobotocore.

**Comment by cecton** [comment]
@asvetlov I think what we could do also is to have an in-between approach: wrap the low-level API of docker-py and make a brand new high-level API.

When I see that kind of PR: https://github.com/docker/docker-py/pull/1657/files I just can't ignore the fact that the API compatibility maintenance is better done by Docker themselves.

On the other hand, the high level API of docker-py is difficult to port to aiohttp. This one could remain unique and maintained by us, this will probably reduce the maintenance.

**Comment by jettify** [comment]
Wrapper is possible only with full cooperation of docker-py folks, otherwise it is pain to support, separate client looks more promising to me.

**Comment by cecton** [comment]
> Wrapper is possible only with full cooperation of docker-py folks

That's a good point! Actually I had recently submitted some changes to allow some of the methods to work more smoothly as coroutines. They were actually already changing the things I wanted to change.

I have worked with the maintainer on another PR (for Docker Compose) and I think if we announce officially as a team that we want to improve the wrapper and, in exchange, improve docker-py. Both sides will see the perks.

> separate client looks more promising to me.

To me also it seems that we can do more and, with time, getting something more polished. But there are 2 reasons for me to think that we may be not on the best track:

1. I'm using for the time being aiodocker-py (the wrapper) on some projects for my work. I would prefer to use aiodocker and I wanted to switch but in the current state of the project it is not stable enough in my opinion. With aiodocker-py I reached more quickly a certain level of stability at the cost of the number of features (I only ported the low-level API).
2. docker-py is quite in a unique situation since it has 2 client APIs: the low-level API and the high one. The low-level API are simple methods, well isolated, easy to understand, easy to maintain and docker-py seems to keep this API. So I believe we can do both: wrap the low-level API and create our own high level API.
